File: python/5_Longest_Palindromic_Substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        dp = [[False] * n for _ in range(n)]
        ans = ""
        for length in range(n): # 0,1,2,3,4
            for start in range(0,n-length,1): # 0,1,2,3,4
                end = start + length # [start,end]
                if length == 0:
                    dp[start][end] =  True
                elif length == 1:
                    dp[start][end] =  (s[start] == s[end])
                else:
                    dp[start][end] =  (s[start] == s[end] and dp[start+1][end-1])
                if dp[start][end] and length + 1 > len(ans):
                    ans = s[start:end+1]
        return ans

File: java/5_Longest_Palindromic_Substring.java
class Solution {
    public String longestPalindrome(String s) {
        int n = s.length();
        boolean[][] dp = new boolean[n][n];
        String ans = "";

        for(int l = 0 ; l < n ; l++){
            for(int index = 0 ; index + l < n ; index++){
                int j = index + l;
                if(l == 0){
                    dp[index][j] = true;
                }
                else if(l == 1){
                    dp[index][j] = (s.charAt(index) == s.charAt(j));
                }
                else{
                    dp[index][j] = (s.charAt(index) == s.charAt(j)) && dp[index+1][j-1];
                }
                if(dp[index][j] && l + 1 > ans.length()){
                    ans = s.substring(index,j+1);
                }
            }
        }
        return ans;
    }
}
